File: infra/sdk/ts/src/identity.ts
import { logger } from "./logger";

export type IdentityKind = "user" | "service" | "platform";

export type Identity = {
  kind: IdentityKind;
  subject: string;
  email: string | null;
  scopes: string[];
  actorChain: string[];
  cnfJkt?: string | null;
  sessionId?: string | null;
  subjectToken?: string | null;
  clientInstance?: string | null;
  clientKind?: string | null;
};

export const identityKey = Symbol("platy.identity");

export type Principal = {
  kind: IdentityKind;
  sub: string;
  email?: string;
  act?: string[];
};

export const principalFromIdentity = (
  identity: Pick<Identity, "kind" | "subject" | "email" | "actorChain">,
): Principal => ({
  kind: identity.kind,
  sub: identity.subject,
  ...(identity.email ? { email: identity.email } : {}),
  ...(identity.actorChain.length > 0 ? { act: identity.actorChain } : {}),
});

export type IdentityExchangeLog = {
  audience: string;
  subject_token_type: string;
  actor_token_type?: string;
  act?: string;
  impersonation?: boolean;
  principal?: Principal;
  scopes?: string[];
  reason?: string;
  status?: number;
};

export const identityExchangeRefused = (fields: IdentityExchangeLog): void => {
  logger.warn("identity_exchange_refused", fields);
};

export const identityExchanged = (fields: IdentityExchangeLog): void => {
  logger.info("identity_exchanged", fields);
};

File: infra/sdk/go/identity/identity.go
package identity

type Principal struct {
	Kind  string   `json:"kind"`
	Sub   string   `json:"sub"`
	Email string   `json:"email,omitempty"`
	Act   []string `json:"act,omitempty"`
}

func NewPrincipal(kind, sub, email string, act []string) Principal {
	p := Principal{Kind: kind, Sub: sub, Email: email}
	if len(act) > 0 {
		p.Act = append([]string(nil), act...)
	}
	return p
}

func PrincipalMap(principal Principal) map[string]any {
	fields := map[string]any{
		"kind": principal.Kind,
		"sub":  principal.Sub,
	}
	if principal.Email != "" {
		fields["email"] = principal.Email
	}
	if len(principal.Act) > 0 {
		fields["act"] = principal.Act
	}
	return fields
}

type ExchangeLog struct {
	Audience         string     `json:"audience"`
	SubjectTokenType string     `json:"subject_token_type"`
	ActorTokenType   string     `json:"actor_token_type,omitempty"`
	Act              string     `json:"act,omitempty"`
	Impersonation    bool       `json:"impersonation,omitempty"`
	Principal        *Principal `json:"principal,omitempty"`
	Scopes           []string   `json:"scopes,omitempty"`
	Reason           string     `json:"reason,omitempty"`
}

func ExchangeRefusedLog(
	audience, subjectTokenType, actorTokenType, act, reason string,
	impersonation bool,
	principal *Principal,
) []any {
	fields := []any{
		"audience", audience,
		"subject_token_type", subjectTokenType,
	}
	if actorTokenType != "" {
		fields = append(fields, "actor_token_type", actorTokenType)
	}
	if act != "" {
		fields = append(fields, "act", act)
	}
	if impersonation {
		fields = append(fields, "impersonation", true)
	}
	if principal != nil {
		fields = append(fields, "principal", PrincipalMap(*principal))
	}
	if reason != "" {
		fields = append(fields, "reason", reason)
	}
	return fields
}

func ExchangedLog(
	audience, subjectTokenType, actorTokenType, act string,
	impersonation bool,
	principal *Principal,
	scopes []string,
) []any {
	fields := ExchangeRefusedLog(audience, subjectTokenType, actorTokenType, act, "", impersonation, principal)
	if len(scopes) > 0 {
		fields = append(fields, "scopes", scopes)
	}
	return fields
}
